### PR TITLE
Clean up transforming some data classes to records

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR cleans up some data classes from the Raft log/journal by transforming them to records. One thing I learned, Kryo does not play nice with records, so if we want to keep rolling updates, we cannot switch anything in the protocol to records. To that end, I moved the `PersistedRaftRecord` into the protocol sub package, since it's used only for passing over the network.

## Related issues

This came out of the PR to refactor the log storage appender tests, as I had to dig a bit into Atomix storage.

related to #9244 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: atomix/cluster/src/main/java/io/atomix/raft/partition/impl/RaftNamespaces.java [hunk 1]
```diff
@@ -83,9 +77,7 @@ public final class RaftNamespaces {
           .register(MemberId.class)
           .register(RaftMember.Type.class)
           .register(Instant.class)
-          .register(Configuration.class)
-          .register(ApplicationEntry.class)
-          .register(RaftLogEntry.class)
+          .nextId(528) // for backwards compatibility
```
Comment (Zelldon) [inline atomix/cluster/src/main/java/io/atomix/raft/partition/impl/RaftNamespaces.java:80]: :upside_down_face: TIL
:question: Why? and do we no longer need these classes?

File: atomix/cluster/src/main/java/io/atomix/raft/protocol/PersistedRaftRecord.java [hunk 2]
```diff
@@ -1,19 +1,11 @@
...
- * See the License for the specific language governing permissions and
- * limitations under the License.
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
  */
-package io.atomix.raft.storage.log;
+package io.atomix.raft.protocol;
```
Comment (Zelldon) [inline atomix/cluster/src/main/java/io/atomix/raft/protocol/PersistedRaftRecord.java:8]: :question: Why was this moved?

File: atomix/cluster/src/main/java/io/atomix/raft/protocol/PersistedRaftRecord.java [hunk 3]
```diff
@@ -1,19 +1,11 @@
...
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
```
Comment (Zelldon) [inline atomix/cluster/src/main/java/io/atomix/raft/protocol/PersistedRaftRecord.java:6]: :x: Why this change? Since when we have here a different License?

File: atomix/cluster/src/main/java/io/atomix/raft/storage/log/IndexedRaftLogEntryImpl.java [hunk 4]
```diff
@@ -16,42 +16,18 @@
...
-import java.util.Objects;
 
 /** Indexed journal entry. */
-class IndexedRaftLogEntryImpl implements IndexedRaftLogEntry {
-
-  private final long index;
-  private final long term;
-  private final RaftEntry entry;
-  private final JournalRecord record;
+record IndexedRaftLogEntryImpl(long index, long term, RaftEntry entry, JournalRecord record)
```
Comment (Zelldon) [inline atomix/cluster/src/main/java/io/atomix/raft/storage/log/IndexedRaftLogEntryImpl.java:25]: :+1:

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]
Thanks @npepinpe I have some comments and questions, before we can merge this. Regarding the License header this seems to be for me a blocker and need more clarification. I like the change to records :+1: 

> This PR cleans up some data classes from the Raft log/journal by transforming them to records. One thing I learned, Kryo does not play nice with records, so if we want to keep rolling updates, we cannot switch anything in the protocol to records.

We don't "support" rolling updates :sweat_smile: 


> To that end, I moved the PersistedRaftRecord into the protocol sub package, since it's used only for passing over the network.

I don't get this change. The name says to me it is for perisitance, why it is not part of the storage :confused: :thinking:

File: atomix/cluster/src/main/java/io/atomix/raft/protocol/PersistedRaftRecord.java (hunk 3 above)
Comment (npepinpe) [reply atomix/cluster/src/main/java/io/atomix/raft/protocol/PersistedRaftRecord.java:6]: Ah, it seems I IntelliJ added it automatically. I thought it only did that on new files, not when moving files...anyway it wasn't on purpose at all, sorry

Comment (Zelldon) [reply atomix/cluster/src/main/java/io/atomix/raft/protocol/PersistedRaftRecord.java:6]: I guess we need to fix this, otherwise, we have the project everywhere different license headers in the modules.

File: atomix/cluster/src/main/java/io/atomix/raft/protocol/PersistedRaftRecord.java (hunk 2 above)
Comment (npepinpe) [reply atomix/cluster/src/main/java/io/atomix/raft/protocol/PersistedRaftRecord.java:16]: Even if it has `Persisted` in the name, this class is a DTO - it's used to transmit `IndexedRaftLogEntry` over the network during replication, then used to create a new one on the receiving side. It's never really used to interact with log itself except in the context of network replication.

File: atomix/cluster/src/main/java/io/atomix/raft/partition/impl/RaftNamespaces.java (hunk 1 above)
Comment (npepinpe) [reply atomix/cluster/src/main/java/io/atomix/raft/partition/impl/RaftNamespaces.java:80]: We do use them internally, but not as part of our protocol. Meaning you can change them as long as the SBE serialization handles them properly, but you don't have to worry about messing up Kryo serialization which is based on the structure of your class (e.g. field ordering and such).

Since changing classes that need to be serialized with Kryo is, uh, error prone (or at least you have to be very careful), I wanted to make it clear which you can and which you cannot change.

File: atomix/cluster/src/main/java/io/atomix/raft/protocol/PersistedRaftRecord.java (hunk 3 above)
Comment (npepinpe) [reply atomix/cluster/src/main/java/io/atomix/raft/protocol/PersistedRaftRecord.java]: I'll look at how this happened. I'm 100% sure I didn't create a new file or copy this manually...I'll try to reproduce it :thinking:

File: atomix/cluster/src/main/java/io/atomix/raft/partition/impl/RaftNamespaces.java (hunk 1 above)
Comment (Zelldon) [reply atomix/cluster/src/main/java/io/atomix/raft/partition/impl/RaftNamespaces.java:80]: Thanks :+1:

Comment (Zelldon) [reply atomix/cluster/src/main/java/io/atomix/raft/partition/impl/RaftNamespaces.java:80]: I guess we can use this trick on other places where I used Void.class :D

Comment (npepinpe) [reply atomix/cluster/src/main/java/io/atomix/raft/partition/impl/RaftNamespaces.java:80]: Oh, I forgot about that, that's also clever and would've been less effort than what I did (manually counting everything :sweat_smile:)

**Review by ChrisKujawa — APPROVED** [review]
